## 문제 06 유형(DataSet_06.csv 이용)

## 주택 관련 정보를 바탕으로 주택 가격을 예측해 보고자 한다. 다음은 확보한 주택 관련 데이터로 총 19개 컬럼으로 구성되어 있다.

In [1]:
import pandas as pd

In [2]:
dataset6 = pd.read_csv('DataSet/DataSet_06.csv')

dataset6

id             date     price  bedrooms  bathrooms  sqft_living  \
0     2597000006  20150309T000000  347500.0         3       1.50         1180   
1     8813400345  20150414T000000  575000.0         2       1.00          980   
2      226059106  20150102T000000  489500.0         3       1.75         2090   
3     8645500270  20141028T000000  246000.0         4       1.75         1720   
4     7010700550  20141114T000000  595000.0         3       2.50         2030   
...          ...              ...       ...       ...        ...          ...   
4318  7663700531  20150106T000000  325000.0         2       1.00          620   
4319  7399200770  20141209T000000  417400.0         3       3.00         2680   
4320  1324300126  20150313T000000  415000.0         2       2.50         1160   
4321  8562750250  20140704T000000  600000.0         3       2.50         2320   
4322  4036800170  20140623T000000  453000.0         4       1.75         2120   

      sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0         8353     1.0           0     0          3      7        1180   
1         3663     1.0           0     0          5      7         980   
2        65558     1.0           0     0          3      8        1330   
3         7455     1.0           0     0          4      7        1020   
4         5100     2.0           0     0          3      7        2030   
...        ...     ...         ...   ...        ...    ...         ...   
4318     14823     1.0           0     0          3      6         620   
4319     12285     1.0           0     0          4      8        2680   
4320      1219     3.0           0     0          3      8        1160   
4321      7609     2.0           0     0          3      8        2320   
4322      7420     1.0           0     0          4      7        1060   

      sqft_basement  yr_built  yr_renovated  zipcode  sqft_living15  \
0                 0      1960             0    98155           1710   
1                 0      1909             0    98105           1620   
2               760      1977             0    98072           2450   
3               700      1963             0    98058           1720   
4                 0      2008             0    98199           1790   
...             ...       ...           ...      ...            ...   
4318              0      1926             0    98125           1400   
4319              0      1970             0    98055           2610   
4320              0      2007             0    98107           1320   
4321              0      2003             0    98027           2590   
4322           1060      1956             0    98008           1540   

      sqft_lot15  
0           8748  
1           3706  
2          47178  
3           7700  
4           4380  
...          ...  
4318        7930  
4319        9558  
4320        2800  
4321        4000  
4322        7420  

[4323 rows x 19 columns]

In [3]:
dataset6.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [4]:
dataset6.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [5]:
dataset6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4323 entries, 0 to 4322
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             4323 non-null   int64  
 1   date           4323 non-null   object 
 2   price          4323 non-null   float64
 3   bedrooms       4323 non-null   int64  
 4   bathrooms      4323 non-null   float64
 5   sqft_living    4323 non-null   int64  
 6   sqft_lot       4323 non-null   int64  
 7   floors         4323 non-null   float64
 8   waterfront     4323 non-null   int64  
 9   view           4323 non-null   int64  
 10  condition      4323 non-null   int64  
 11  grade          4323 non-null   int64  
 12  sqft_above     4323 non-null   int64  
 13  sqft_basement  4323 non-null   int64  
 14  yr_built       4323 non-null   int64  
 15  yr_renovated   4323 non-null   int64  
 16  zipcode        4323 non-null   int64  
 17  sqft_living15  4323 non-null   int64  
 18  sqft_lot

### 1. 강변 조망이 가능한지 여부(waterfront)에 따라 평균 주택 가격을 계산하고 조망이 가능한 경우와 그렇지 않은 경우의 평균 가격 차이의 절대값을 구하시오. 답은 소수점 이하는 버리고 정수부만 기술하시오.<br>(답안 예시) 1234567

In [6]:
Q1 = dataset6.groupby('waterfront')['price'].mean()

Q1

waterfront
0    5.321651e+05
1    1.699438e+06
Name: price, dtype: float64

In [7]:
abs(Q1[0] - Q1[1])

1167272.748018648

#### 정답 : ***1167272***

In [8]:
# 수업 풀이
q1_1 = dataset6[dataset6.waterfront==1]['price'].mean()
q1_2 = dataset6[dataset6.waterfront==0]['price'].mean()

In [9]:
abs(q1_1 - q1_2)

1167272.748018648

### 2. price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, yr_built 등 7개의 변수 간의 상관분석을 수행하고 price와의 상관계수의 절대값이 가장 큰 변수와 가장 작은 변수를 차례로 기술하시오.<br>(답안 예시) view, zipcode

In [10]:
Q2 = dataset6[['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'yr_built']]

Q2

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  yr_built
0     347500.0         3       1.50         1180      8353     1.0      1960
1     575000.0         2       1.00          980      3663     1.0      1909
2     489500.0         3       1.75         2090     65558     1.0      1977
3     246000.0         4       1.75         1720      7455     1.0      1963
4     595000.0         3       2.50         2030      5100     2.0      2008
...        ...       ...        ...          ...       ...     ...       ...
4318  325000.0         2       1.00          620     14823     1.0      1926
4319  417400.0         3       3.00         2680     12285     1.0      1970
4320  415000.0         2       2.50         1160      1219     3.0      2007
4321  600000.0         3       2.50         2320      7609     2.0      2003
4322  453000.0         4       1.75         2120      7420     1.0      1956

[4323 rows x 7 columns]

In [11]:
Q2.corr()['price'][1:].abs().nlargest(2)

sqft_living    0.709157
bathrooms      0.534277
Name: price, dtype: float64

#### 정답 : ***sqft_living, bathrooms*** → 문제 잘못 읽음...!

In [12]:
# 수업 풀이
x_var = ['price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'yr_built']

In [13]:
q2 = dataset6[x_var]

In [14]:
q2_out = q2.corr().iloc[1:, 0]

q2_out

bedrooms       0.279824
bathrooms      0.534277
sqft_living    0.709157
sqft_lot       0.106341
floors         0.270423
yr_built       0.059895
Name: price, dtype: float64

In [15]:
print(q2_out.max())
print(q2_out.min())

0.7091565835706303
0.059895410779646746


In [16]:
print(q2_out.nlargest(1))
print(q2_out.nsmallest(1))

sqft_living    0.709157
Name: price, dtype: float64
yr_built    0.059895
Name: price, dtype: float64


#### 정답 : ***sqft_living, yr_built***

### 3. id, date, 그리고 zipcode를 제외한 모든 변수를 독립변수로, price를 종속변수로 하여 회귀분석을 수행하시오. 통계적 유의성을 갖지 못하는 독립변수를 제거하면 회귀 모형에 남는 변수는 모두 몇 개인가? 이 때 음의 회귀계수를 가지는 변수는 몇 개인가?<br>(답안 예시) 5, 3

In [17]:
from sklearn.linear_model import LinearRegression
from statsmodels.formula.api import ols

In [18]:
Q3 = dataset6.drop(['id', 'date', 'zipcode'], axis=1)

Q3

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0     347500.0         3       1.50         1180      8353     1.0   
1     575000.0         2       1.00          980      3663     1.0   
2     489500.0         3       1.75         2090     65558     1.0   
3     246000.0         4       1.75         1720      7455     1.0   
4     595000.0         3       2.50         2030      5100     2.0   
...        ...       ...        ...          ...       ...     ...   
4318  325000.0         2       1.00          620     14823     1.0   
4319  417400.0         3       3.00         2680     12285     1.0   
4320  415000.0         2       2.50         1160      1219     3.0   
4321  600000.0         3       2.50         2320      7609     2.0   
4322  453000.0         4       1.75         2120      7420     1.0   

      waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0              0     0          3      7        1180              0      1960   
1              0     0          5      7         980              0      1909   
2              0     0          3      8        1330            760      1977   
3              0     0          4      7        1020            700      1963   
4              0     0          3      7        2030              0      2008   
...          ...   ...        ...    ...         ...            ...       ...   
4318           0     0          3      6         620              0      1926   
4319           0     0          4      8        2680              0      1970   
4320           0     0          3      8        1160              0      2007   
4321           0     0          3      8        2320              0      2003   
4322           0     0          4      7        1060           1060      1956   

      yr_renovated  sqft_living15  sqft_lot15  
0                0           1710        8748  
1                0           1620        3706  
2                0           2450       47178  
3                0           1720        7700  
4                0           1790        4380  
...            ...            ...         ...  
4318             0           1400        7930  
4319             0           2610        9558  
4320             0           1320        2800  
4321             0           2590        4000  
4322             0           1540        7420  

[4323 rows x 16 columns]

In [19]:
X = Q3.drop('price', axis=1)
y = Q3.price

In [20]:
lm = LinearRegression(fit_intercept=True)
lm.fit(X, y)

LinearRegression()

In [21]:
lm.coef_

array([-2.49955251e+04,  5.17923445e+04,  9.23080441e+01,  5.44484390e-03,
        3.74819905e+04,  6.27652844e+05,  4.03977832e+04,  1.57057341e+04,
        1.30627396e+05,  3.37357935e+01,  5.85722507e+01, -3.66689567e+03,
        1.77031918e+01,  2.30114429e+01, -2.32290365e-01])

In [22]:
var_list = Q3.columns[1:]

var_list

Index(['bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors',
       'waterfront', 'view', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'sqft_living15',
       'sqft_lot15'],
      dtype='object')

In [23]:
form = 'price~'+'+'.join(var_list)

In [24]:
ols1 = ols(form, data=Q3).fit()

ols1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.681
Model:                            OLS   Adj. R-squared:                  0.680
Method:                 Least Squares   F-statistic:                     656.0
Date:                Tue, 24 Aug 2021   Prob (F-statistic):               0.00
Time:                        15:10:13   Log-Likelihood:                -58960.
No. Observations:                4323   AIC:                         1.180e+05
Df Residuals:                    4308   BIC:                         1.180e+05
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept      6.302e+06   2.95e+05     21.401      0.000    5.73e+06    6.88e+06
bedrooms        -2.5e+04   3680.510     -6.791      0.000   -3.22e+04   -1.78e+04
bathrooms      5.179e+04   7163.408      7.230      0.000    3.77e+04    6.58e+04
sqft_living      92.3080      5.074     18.192      0.000      82.360     102.256
sqft_lot          0.0054      0.093      0.059      0.953      -0.176       0.187
floors         3.748e+04   8141.392      4.604      0.000    2.15e+04    5.34e+04
waterfront     6.277e+05   3.91e+04     16.064      0.000    5.51e+05    7.04e+05
view            4.04e+04   4797.344      8.421      0.000     3.1e+04    4.98e+04
condition      1.571e+04   5317.387      2.954      0.003    5280.918    2.61e+04
grade          1.306e+05   4700.604     27.789      0.000    1.21e+05     1.4e+05
sqft_above       33.7358      4.981      6.773      0.000      23.971      43.501
sqft_basement    58.5723      5.840     10.030      0.000      47.123      70.021
yr_built      -3666.8957    150.805    -24.316      0.000   -3962.551   -3371.241
yr_renovated     17.7032      8.265      2.142      0.032       1.499      33.908
sqft_living15    23.0114      7.561      3.044      0.002       8.188      37.834
sqft_lot15       -0.2323      0.142     -1.637      0.102      -0.511       0.046
==============================================================================
Omnibus:                     2170.721   Durbin-Watson:                   2.011
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            38362.875
Skew:                           1.971   Prob(JB):                         0.00
Kurtosis:                      17.051   Cond. No.                     3.23e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.56e-22. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

#### 정답 : ***13, 3***

In [25]:
# 수업 풀이
q3 = dataset6.drop(columns=['id', 'date', 'zipcode'], axis=1)

In [26]:
x_var = q3.columns.drop('price')

In [27]:
from statsmodels.formula.api import ols

In [28]:
form = 'price~'+'+'.join(x_var)

form

'price~bedrooms+bathrooms+sqft_living+sqft_lot+floors+waterfront+view+condition+grade+sqft_above+sqft_basement+yr_built+yr_renovated+sqft_living15+sqft_lot15'

In [29]:
ols2 = ols(form, data=q3).fit()

ols2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.681
Model:                            OLS   Adj. R-squared:                  0.680
Method:                 Least Squares   F-statistic:                     656.0
Date:                Tue, 24 Aug 2021   Prob (F-statistic):               0.00
Time:                        15:10:13   Log-Likelihood:                -58960.
No. Observations:                4323   AIC:                         1.180e+05
Df Residuals:                    4308   BIC:                         1.180e+05
Df Model:                          14                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept      6.302e+06   2.95e+05     21.401      0.000    5.73e+06    6.88e+06
bedrooms        -2.5e+04   3680.510     -6.791      0.000   -3.22e+04   -1.78e+04
bathrooms      5.179e+04   7163.408      7.230      0.000    3.77e+04    6.58e+04
sqft_living      92.3080      5.074     18.192      0.000      82.360     102.256
sqft_lot          0.0054      0.093      0.059      0.953      -0.176       0.187
floors         3.748e+04   8141.392      4.604      0.000    2.15e+04    5.34e+04
waterfront     6.277e+05   3.91e+04     16.064      0.000    5.51e+05    7.04e+05
view            4.04e+04   4797.344      8.421      0.000     3.1e+04    4.98e+04
condition      1.571e+04   5317.387      2.954      0.003    5280.918    2.61e+04
grade          1.306e+05   4700.604     27.789      0.000    1.21e+05     1.4e+05
sqft_above       33.7358      4.981      6.773      0.000      23.971      43.501
sqft_basement    58.5723      5.840     10.030      0.000      47.123      70.021
yr_built      -3666.8957    150.805    -24.316      0.000   -3962.551   -3371.241
yr_renovated     17.7032      8.265      2.142      0.032       1.499      33.908
sqft_living15    23.0114      7.561      3.044      0.002       8.188      37.834
sqft_lot15       -0.2323      0.142     -1.637      0.102      -0.511       0.046
==============================================================================
Omnibus:                     2170.721   Durbin-Watson:                   2.011
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            38362.875
Skew:                           1.971   Prob(JB):                         0.00
Kurtosis:                      17.051   Cond. No.                     3.23e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.56e-22. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [30]:
len(x_var)

15

In [31]:
sum((ols2.pvalues<0.05)) - 1   # intercept 제외

13

In [32]:
ols2.pvalues<0.05

Intercept         True
bedrooms          True
bathrooms         True
sqft_living       True
sqft_lot         False
floors            True
waterfront        True
view              True
condition         True
grade             True
sqft_above        True
sqft_basement     True
yr_built          True
yr_renovated      True
sqft_living15     True
sqft_lot15       False
dtype: bool

## 문제 07 유형(DataSet_07.csv 이용)

## 대학원 진학을 위하여 어떤 항목이 중요하게 영향을 미치는지 아래 데이터로 분석하고자 한다.

In [33]:
import pandas as pd

In [34]:
dataset7 = pd.read_csv('DataSet/DataSet_07.csv')

dataset7

Serial_No  GRE  TOEFL  University_Rating  SOP  LOR  CGPA  Research  \
0            1  337    118                  4  4.5  4.5  9.65         1   
1            2  324    107                  4  4.0  4.5  8.87         1   
2            3  316    104                  3  3.0  3.5  8.00         1   
3            4  322    110                  3  3.5  2.5  8.67         1   
4            5  314    103                  2  2.0  3.0  8.21         0   
..         ...  ...    ...                ...  ...  ...   ...       ...   
395        396  324    110                  3  3.5  3.5  9.04         1   
396        397  325    107                  3  3.0  3.5  9.11         1   
397        398  330    116                  4  5.0  4.5  9.45         1   
398        399  312    103                  3  3.5  4.0  8.78         0   
399        400  333    117                  4  5.0  4.0  9.66         1   

     Chance_of_Admit  
0               0.92  
1               0.76  
2               0.72  
3               0.80  
4               0.65  
..               ...  
395             0.82  
396             0.84  
397             0.91  
398             0.67  
399             0.95  

[400 rows x 9 columns]

In [35]:
dataset7.columns

Index(['Serial_No', 'GRE', 'TOEFL', 'University_Rating', 'SOP', 'LOR', 'CGPA',
       'Research', 'Chance_of_Admit'],
      dtype='object')

In [36]:
dataset7.dtypes

Serial_No              int64
GRE                    int64
TOEFL                  int64
University_Rating      int64
SOP                  float64
LOR                  float64
CGPA                 float64
Research               int64
Chance_of_Admit      float64
dtype: object

In [37]:
dataset7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Serial_No          400 non-null    int64  
 1   GRE                400 non-null    int64  
 2   TOEFL              400 non-null    int64  
 3   University_Rating  400 non-null    int64  
 4   SOP                400 non-null    float64
 5   LOR                400 non-null    float64
 6   CGPA               400 non-null    float64
 7   Research           400 non-null    int64  
 8   Chance_of_Admit    400 non-null    float64
dtypes: float64(4), int64(5)
memory usage: 28.2 KB


### 1. 합격 가능성에 GRE, TOEFL, CGPA 점수 가운데 가장 영향이 큰 것이 어떤 점수인지 알아보기 위해서 상관 분석을 수행한다.<br>- 피어슨(Pearson) 상관계수 값을 구한다.<br>- Chance_of_Admit와의 가장 큰 상관계수 값을 가지는 항목의 상관계수를 소수점 넷째 자리에서 반올림하여 셋째 자리까지 기술하시오.<br>(답안 예시) 0.123

In [38]:
Q1 = dataset7[['GRE', 'TOEFL', 'CGPA', 'Chance_of_Admit']]

Q1.corr()['Chance_of_Admit'][:-1].max().round(3)

0.873

#### 정답 : ***0.873***

In [39]:
# 수업 풀이
q1 = dataset7[['GRE', 'TOEFL', 'CGPA', 'Chance_of_Admit']].corr()

q1.iloc[:-1, -1].max().round(3)

0.873

### 2. GRE 점수의 평균 이상을 받은 그룹과 평균 미만을 받은 그룹의 CGPA 평균은 차이가 있는지 검정을 하고자 한다.<br>- 적절한 검정 방법을 선택하고 양측 검정을 수행하시오 (등분산으로 가정)<br>- 검정 결과, 검정통계량의 추정치를 소수점 셋째 자리에서 반올림하여 소수점 두 자리까지 기술하시오.<br>(답안 예시) 1.23

In [40]:
Q2 = dataset7[['GRE', 'CGPA']]

In [41]:
from scipy.stats import ttest_ind

In [42]:
Q2_out = ttest_ind(Q2[Q2.GRE>=Q2.GRE.mean()].CGPA,
                   Q2[Q2.GRE<Q2.GRE.mean()].CGPA,
                   equal_var=True)   # 등분산

Q2_out[0].round(2)

19.44

#### 정답 : ***19.44***

In [43]:
# 수업 풀이
q2 = dataset7.copy()

In [44]:
GRE_avg = q2.GRE.mean()

GRE_avg

316.8075

In [45]:
q2['GRE_gr'] = 'under'
q2.loc[q2['GRE']>=GRE_avg, 'GRE_gr'] = 'over'

In [46]:
from scipy.stats import ttest_ind

In [47]:
a = q2[q2['GRE_gr']=='over']['CGPA']
b = q2[q2['GRE_gr']=='under']['CGPA']

ttest_ind(a, b, equal_var=True)   # equal_var의 default는 True

Ttest_indResult(statistic=19.443291692470982, pvalue=1.1127037052806161e-59)

### 3. Chance_of_Admit 확률이 0.5를 초과하면 합격으로, 이하이면 불합격으로 구분하고 로지스틱 회귀분석을 수행하시오.<br>- 원데이터만 사용하고, 원데이터 가운데 Serial_No와 Label은 모형에서 제외<br>- 각 설정값은 다음과 같이 지정하고, 언급되지 않은 사항은 기본 설정값을 사용하시오. (Seed : 123)<br>- 로지스틱 회귀분석 수행 결과에서 로지스틱 회귀계수의 절대값이 가장 큰 변수와 그 값을 기술하시오. (로지스틱 회귀계수는 반올림하여 소수점 둘째 자리까지 / Intercept는 제외)<br>(답안 예시) abc, 0.12

In [48]:
# 수업 풀이
q3 = dataset7.drop(columns=['Serial_No'], axis=1)

In [49]:
import numpy as np

q3['ch_ad'] = np.where(q3.Chance_of_Admit>0.5, 1, 0)

q3

GRE  TOEFL  University_Rating  SOP  LOR  CGPA  Research  Chance_of_Admit  \
0    337    118                  4  4.5  4.5  9.65         1             0.92   
1    324    107                  4  4.0  4.5  8.87         1             0.76   
2    316    104                  3  3.0  3.5  8.00         1             0.72   
3    322    110                  3  3.5  2.5  8.67         1             0.80   
4    314    103                  2  2.0  3.0  8.21         0             0.65   
..   ...    ...                ...  ...  ...   ...       ...              ...   
395  324    110                  3  3.5  3.5  9.04         1             0.82   
396  325    107                  3  3.0  3.5  9.11         1             0.84   
397  330    116                  4  5.0  4.5  9.45         1             0.91   
398  312    103                  3  3.5  4.0  8.78         0             0.67   
399  333    117                  4  5.0  4.0  9.66         1             0.95   

     ch_ad  
0        1  
1        1  
2        1  
3        1  
4        1  
..     ...  
395      1  
396      1  
397      1  
398      1  
399      1  

[400 rows x 9 columns]

In [50]:
from sklearn.linear_model import LogisticRegression

In [51]:
X = q3.drop(columns=['Chance_of_Admit', 'ch_ad'])
y = q3.ch_ad

log_reg = LogisticRegression(fit_intercept=True,
                             random_state=12,
                             solver='liblinear').fit(X, y)

In [52]:
abs(log_reg.coef_)

array([[0.08778604, 0.10686963, 0.05594224, 0.23228922, 0.98336396,
        1.97858509, 0.66263161]])

In [53]:
q3_out = pd.DataFrame({'var': X.columns.values,
                       'coef': abs(log_reg.coef_).reshape(-1)})

q3_out

var      coef
0                GRE  0.087786
1              TOEFL  0.106870
2  University_Rating  0.055942
3                SOP  0.232289
4                LOR  0.983364
5               CGPA  1.978585
6           Research  0.662632

#### 정답 : ***CGPA, 1.98***

## 문제 08 유형(DataSet_08.csv 이용)

## 스타트업 기업들의 수익성에 대한 분석을 하기 위하여 아래와 같은 데이터를 입수하였다.

In [54]:
import pandas as pd

In [55]:
dataset8 = pd.read_csv('DataSet/DataSet_08.csv')

dataset8

RandD_Spend  Administration  Marketing_Spend       State     Profit
0     165349.20       136897.80        471784.10    New York  192261.83
1     162597.70       151377.59        443898.53  California  191792.06
2     153441.51       101145.55        407934.54     Florida  191050.39
3     144372.41       118671.85        383199.62    New York  182901.99
4     142107.34        91391.77        366168.42     Florida  166187.94
5     131876.90        99814.71        362861.36    New York  156991.12
6     134615.46       147198.87        127716.82  California  156122.51
7     130298.13       145530.06        323876.68     Florida  155752.60
8     120542.52       148718.95        311613.29    New York  152211.77
9     123334.88       108679.17        304981.62  California  149759.96
10    101913.08       110594.11        229160.95     Florida  146121.95
11    100671.96        91790.61        249744.55  California  144259.40
12     93863.75       127320.38        249839.44     Florida  141585.52
13     91992.39       135495.07        252664.93  California  134307.35
14    119943.24       156547.42        256512.92     Florida  132602.65
15    114523.61       122616.84        261776.23    New York  129917.04
16     78013.11       121597.55        264346.06  California  126992.93
17     94657.16       145077.58        282574.31    New York  125370.37
18     91749.16       114175.79        294919.57     Florida  124266.90
19     86419.70       153514.11             0.00    New York  122776.86
20     76253.86       113867.30        298664.47  California  118474.03
21     78389.47       153773.43        299737.29    New York  111313.02
22     73994.56       122782.75        303319.26     Florida  110352.25
23     67532.53       105751.03        304768.73     Florida  108733.99
24     77044.01        99281.34        140574.81    New York  108552.04
25     64664.71       139553.16        137962.62  California  107404.34
26     75328.87       144135.98        134050.07     Florida  105733.54
27     72107.60       127864.55        353183.81    New York  105008.31
28     66051.52       182645.56        118148.20     Florida  103282.38
29     65605.48       153032.06        107138.38    New York  101004.64
30     61994.48       115641.28         91131.24     Florida   99937.59
31     61136.38       152701.92         88218.23    New York   97483.56
32     63408.86       129219.61         46085.25  California   97427.84
33     55493.95       103057.49        214634.81     Florida   96778.92
34     46426.07       157693.92        210797.67  California   96712.80
35     46014.02        85047.44        205517.64    New York   96479.51
36     28663.76       127056.21        201126.82     Florida   90708.19
37     44069.95        51283.14        197029.42  California   89949.14
38     20229.59        65947.93        185265.10    New York   81229.06
39     38558.51        82982.09        174999.30  California   81005.76
40     28754.33       118546.05        172795.67  California   78239.91
41     27892.92        84710.77        164470.71     Florida   77798.83
42     23640.93        96189.63        148001.11  California   71498.49
43     15505.73       127382.30         35534.17    New York   69758.98
44     22177.74       154806.14         28334.72  California   65200.33
45      1000.23       124153.04          1903.93    New York   64926.08
46      1315.46       115816.21        297114.46     Florida   49490.75
47         0.00       135426.92             0.00  California   42559.73
48       542.05        51743.15             0.00    New York   35673.41
49         0.00       116983.80         45173.06  California   14681.40

In [56]:
dataset8.columns

Index(['RandD_Spend', 'Administration', 'Marketing_Spend', 'State', 'Profit'], dtype='object')

In [57]:
dataset8.dtypes

RandD_Spend        float64
Administration     float64
Marketing_Spend    float64
State               object
Profit             float64
dtype: object

In [58]:
dataset8.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RandD_Spend      50 non-null     float64
 1   Administration   50 non-null     float64
 2   Marketing_Spend  50 non-null     float64
 3   State            50 non-null     object 
 4   Profit           50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


### 1. 각 주(State)별 데이터 구성비를 소수점 둘째 자리까지 구하고, 알파벳 순으로 기술하시오(주 이름 기준).<br>(답안 예시) 0.12, 0.34, 0.54

### 2. 주별 이익의 평균을 구하고, 평균 이익이 가장 큰 주와 작은 주의 차이를 구하시오. 차이값은 소수점 이하는 버리고 정수 부분만 기술하시오.<br>(답안 예시) 1234

### 3. 독립변수로 RandD_Spend, Administration, Marketing_Spend를 사용하여 Profit을 주별로 예측하는 회귀 모형을 만들고, 이 회귀 모형을 사용하여 학습오차를 산출하시오.<br>- 주별로 계산된 학습오차 중 MAPE 기준으로 가장 낮은 오차를 보이는 주는 어느 주이고 그 값은 무엇인가? (반올림하여 소수점 둘째 자리까지 기술하시오)<br>- (MAPE = Σ ( | y - y ̂ | / y ) * 100/n )<br>(답안 예시) ABC, 1.56

In [59]:
from sklearn.linear_model import LinearRegression

In [60]:
Q3 = dataset8.copy()

Q3.State.unique()

array(['New York', 'California', 'Florida'], dtype=object)

In [61]:
# New York
x_var = ['RandD_Spend', 'Administration', 'Marketing_Spend']
temp = Q3[Q3.State=='New York']

In [62]:
lm_NY = LinearRegression().fit(temp[x_var], temp.Profit)
pred_NY = lm_NY.predict(temp[x_var])

In [63]:
mape_NY = ((abs(temp.Profit - pred_NY) / temp.Profit).sum() * 100/len(temp)).round(2)

mape_NY

8.52

In [64]:
# California
x_var = ['RandD_Spend', 'Administration', 'Marketing_Spend']
temp = Q3[Q3.State=='California']

In [65]:
lm_Ca = LinearRegression().fit(temp[x_var], temp.Profit)
pred_Ca = lm_Ca.predict(temp[x_var])

In [66]:
mape_Ca = ((abs(temp.Profit - pred_Ca) / temp.Profit).sum() * 100/len(temp)).round(2)

mape_Ca

15.23

In [67]:
# Florida
x_var = ['RandD_Spend', 'Administration', 'Marketing_Spend']
temp = Q3[Q3.State=='Florida']

In [68]:
lm_Fl = LinearRegression().fit(temp[x_var], temp.Profit)
pred_Fl = lm_Fl.predict(temp[x_var])

In [69]:
mape_Fl = ((abs(temp.Profit - pred_Fl) / temp.Profit).sum() * 100/len(temp)).round(2)

mape_Fl

5.71

In [70]:
[mape_NY, mape_Ca, mape_Fl]

[8.52, 15.23, 5.71]

#### 정답 : ***Florida, 5.71***

## 문제 09 유형(DataSet_09.csv 이용)

## 항공사에서 고객만족도 조사를 하고 서비스 개선에 활용하고자 아래와 같은 데이터를 준비하였다.

### 1. 데이터 타입을 위 표에 정의된 타입으로 전처리를 한 후, 데이터 파일 내에 결측값은 총 몇 개인가?<br>(답안 예시) 1

### 2. 다음에 제시된 데이터 처리를 하고 카이제곱 독립성 검정을 수행하시오.<br>- 결측값이 있다면 해당 행을 제거하시오.<br>- 나이는 20 이하이면 10, 30 이하이면 20, 40 이하이면 30, 50 이하이면 40, 60 이하이면 50, 60 초과는 60으로 변환하여 Age_gr으로 파생변수를 생성하시오.<br>- Age_gr, Gender, Customer_Type, Class 변수가 satisfaction에 영향이 있는지 카이제곱 독립성 검정을 수행하시오.<br>- 연관성이 있는 것으로 파악된 변수의 검정통계량 추정치를 정수 부분만 기술하시오.<br>(답안 예시) 123

### 3. 고객 만족도를 라벨로 하여 다음과 같이 로지스틱 회귀분석을 수행하시오.<br>- 결측치가 포함된 행은 제거<br>- 데이터를 7대 3으로 분리 (Seed = 123)<br>- 아래의 11개 변수를 Feature로 사용<br>Flight_Distance, Seat_comfort, Food_and_drink, Inflight_wifi_service, Inflight_entertainment,Onboard_service, Leg_room_service, Baggage_handling, Cleanliness, Departure_Delay_in_Minutes, Arrival_Delay_in_Minutes<br><br>- Seed = 123, 이외의 항목은 모두 Default 사용<br>- 예측 정확도를 측정하고 dissatisfied의 f1 score를 소수점 넷째 자리에서 반올림하여 소수점 셋째 자리까지 기술하시오.<br>(답안 예시) 0.123

## 문제 10 유형(DataSet_10.csv 이용)

## 중고 자동차 가격에 대한 분석을 위하여 아래와 같은 데이터를 확보하였다.

In [71]:
import pandas as pd

In [72]:
dataset10 = pd.read_csv('DataSet/DataSet_10.csv')

dataset10

model  engine_power  age_in_days      km  previous_owners  price  \
0     lounge            51          882   25000                1   8900   
1        pop            51         1186   32500                1   8800   
2      sport            74         4658  142228                1   4200   
3     lounge            51         2739  160000                1   6000   
4        pop            73         3074  106880                1   5700   
...      ...           ...          ...     ...              ...    ...   
1533   sport            51         3712  115280                1   5200   
1534  lounge            74         3835  112000                1   4600   
1535     pop            51         2223   60457                1   7500   
1536  lounge            51         2557   80750                1   5990   
1537     pop            51         1766   54276                1   7900   

      Unnamed: 6  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  \
0            NaN         NaN         NaN         NaN          NaN   
1            NaN         NaN         NaN         NaN          NaN   
2            NaN         NaN         NaN         NaN          NaN   
3            NaN         NaN         NaN         NaN          NaN   
4            NaN         NaN         NaN         NaN          NaN   
...          ...         ...         ...         ...          ...   
1533         NaN         NaN         NaN         NaN          NaN   
1534         NaN         NaN         NaN         NaN          NaN   
1535         NaN         NaN         NaN         NaN          NaN   
1536         NaN         NaN         NaN         NaN          NaN   
1537         NaN         NaN         NaN         NaN          NaN   

      Unnamed: 11  Unnamed: 12  Unnamed: 13  Unnamed: 14  Unnamed: 15  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN   
3             NaN          NaN          NaN          NaN          NaN   
4             NaN          NaN          NaN          NaN          NaN   
...           ...          ...          ...          ...          ...   
1533          NaN          NaN          NaN          NaN          NaN   
1534          NaN          NaN          NaN          NaN          NaN   
1535          NaN          NaN          NaN          NaN          NaN   
1536          NaN          NaN          NaN          NaN          NaN   
1537          NaN          NaN          NaN          NaN          NaN   

      Unnamed: 16  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
...           ...  
1533          NaN  
1534          NaN  
1535          NaN  
1536          NaN  
1537          NaN  

[1538 rows x 17 columns]

In [73]:
dataset10.columns

Index(['model', 'engine_power', 'age_in_days', 'km', 'previous_owners',
       'price', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 9',
       'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 12', 'Unnamed: 13',
       'Unnamed: 14', 'Unnamed: 15', 'Unnamed: 16'],
      dtype='object')

In [74]:
dataset10.dtypes

model               object
engine_power         int64
age_in_days          int64
km                   int64
previous_owners      int64
price                int64
Unnamed: 6         float64
Unnamed: 7         float64
Unnamed: 8         float64
Unnamed: 9         float64
Unnamed: 10        float64
Unnamed: 11        float64
Unnamed: 12        float64
Unnamed: 13        float64
Unnamed: 14        float64
Unnamed: 15        float64
Unnamed: 16        float64
dtype: object

In [75]:
dataset10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1538 entries, 0 to 1537
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   model            1538 non-null   object 
 1   engine_power     1538 non-null   int64  
 2   age_in_days      1538 non-null   int64  
 3   km               1538 non-null   int64  
 4   previous_owners  1538 non-null   int64  
 5   price            1538 non-null   int64  
 6   Unnamed: 6       0 non-null      float64
 7   Unnamed: 7       0 non-null      float64
 8   Unnamed: 8       0 non-null      float64
 9   Unnamed: 9       0 non-null      float64
 10  Unnamed: 10      0 non-null      float64
 11  Unnamed: 11      0 non-null      float64
 12  Unnamed: 12      0 non-null      float64
 13  Unnamed: 13      0 non-null      float64
 14  Unnamed: 14      0 non-null      float64
 15  Unnamed: 15      0 non-null      float64
 16  Unnamed: 16      0 non-null      float64
dtypes: float64(11)

### 1. 이전 소유자 수가 한 명이고 엔진 파워가 51인 차에 대해 모델별 하루 평균 운행 거리를 산출하였을 때 가장 낮은 값을 가진 모델이 가장 큰 값을 가진 모델에 대한 비율은 얼마인가? 소수점 셋째 자리에서 반올림하여 소수점 둘째 자리까지 기술하시오. (모델별 평균 → 일평균 → 최대최소 비율 계산)<br>(답안 예시) 0.12

In [76]:
data1 = dataset10.dropna(axis=1, how='all')

data1

model  engine_power  age_in_days      km  previous_owners  price
0     lounge            51          882   25000                1   8900
1        pop            51         1186   32500                1   8800
2      sport            74         4658  142228                1   4200
3     lounge            51         2739  160000                1   6000
4        pop            73         3074  106880                1   5700
...      ...           ...          ...     ...              ...    ...
1533   sport            51         3712  115280                1   5200
1534  lounge            74         3835  112000                1   4600
1535     pop            51         2223   60457                1   7500
1536  lounge            51         2557   80750                1   5990
1537     pop            51         1766   54276                1   7900

[1538 rows x 6 columns]

In [77]:
data1.columns

Index(['model', 'engine_power', 'age_in_days', 'km', 'previous_owners',
       'price'],
      dtype='object')

In [78]:
q1_1 = data1[data1.previous_owners==1]
q1_2 = q1_1[q1_1.engine_power==51]

In [79]:
print(len(data1))
print(len(q1_1))
print(len(q1_2))

1538
1389
1312


In [80]:
q1_3 = pd.pivot_table(data=q1_2, index='model',
                      values=['age_in_days', 'km'])

q1_3

age_in_days            km
model                            
lounge  1343.450366  44123.670846
pop     1800.509804  58097.532680
sport   3031.387755  96460.979592

In [81]:
q1_3['days'] = q1_3['km'] / q1_3['age_in_days']

q1_3

age_in_days            km       days
model                                       
lounge  1343.450366  44123.670846  32.843544
pop     1800.509804  58097.532680  32.267268
sport   3031.387755  96460.979592  31.820733

In [82]:
(q1_3['days'].min() / q1_3['days'].max()).round(2)

0.97

#### 정답 : ***0.97***

### 2. 운행 일수에 대한 운행 거리를 산출하고, 위 1번 문제에서 가장 큰 값을 가지고 있던 모델과 가장 작은 값을 가지고 있던 모델 간의 운행 일수 대비 운행거리 평균이 다른지 적절한 통계 검정을 수행하고 p-value를 소수점 세 자리 이하는 버리고 소수점 두 자리까지 기술하고 기각 여부를 Y / N로 답하시오. (등분산을 가정하고 equal_var = True / var.equal = T로 분석을 실행하시오.)<br>(답안 예시) 0.23, Y

In [83]:
q2 = q1_1.copy()

q2['km_per_day'] = q2['km'] / q2['age_in_days']

q2

model  engine_power  age_in_days      km  previous_owners  price  \
0     lounge            51          882   25000                1   8900   
1        pop            51         1186   32500                1   8800   
2      sport            74         4658  142228                1   4200   
3     lounge            51         2739  160000                1   6000   
4        pop            73         3074  106880                1   5700   
...      ...           ...          ...     ...              ...    ...   
1533   sport            51         3712  115280                1   5200   
1534  lounge            74         3835  112000                1   4600   
1535     pop            51         2223   60457                1   7500   
1536  lounge            51         2557   80750                1   5990   
1537     pop            51         1766   54276                1   7900   

      km_per_day  
0      28.344671  
1      27.403035  
2      30.534135  
3      58.415480  
4      34.769031  
...          ...  
1533   31.056034  
1534   29.204694  
1535   27.196131  
1536   31.579977  
1537   30.733862  

[1389 rows x 7 columns]

In [84]:
big = q2[q2.model=='lounge']['km_per_day']
small = q2[q2.model=='sport']['km_per_day']

In [85]:
from scipy.stats import ttest_ind

In [86]:
ttest_ind(big, small, equal_var=True)

Ttest_indResult(statistic=1.1025375913936746, pvalue=0.2704782824984852)

#### 정답 : ***0.27, N***

### 3. 독립변수로 engine_power, age_in_days, km를 사용하고 종속변수로 price를 사용하여 모델별 선형회귀분석을 수행하고, 산출된 모형을 사용하여 다음과 같은 조건중고차에 대한 가격을 예측하고 예측된 가격을 정수부만 기술하시오.<br>- model : pop / engine_power : 51 / age_in_days : 400 / km : 9500 / previous_owners : 2<br>(답안 예시) 12345

In [87]:
q3 = q1_1.copy()

In [88]:
from sklearn.linear_model import LinearRegression

In [89]:
x_var = ['engine_power', 'age_in_days', 'km']

In [90]:
lm = LinearRegression().fit(q3[x_var], q3.price)

In [91]:
new_data = pd.DataFrame({
    'engine_power': [51],
    'age_in_days': [400],
    'km': [9500]
})

new_data

engine_power  age_in_days    km
0            51          400  9500

In [92]:
new_data2 = np.array([[51, 400, 9500]])

new_data2

array([[  51,  400, 9500]])

In [93]:
lm.predict(new_data)

array([10463.49541741])

In [94]:
lm.predict(new_data2)

array([10463.49541741])

#### 정답 : ***10463***